docs/migrating.rst

```diff
@@ -0,0 +1,235 @@
+
+There are some methods from the ``tasklet`` module that are not implemented in
+this version of `ndb`, mainly because of changes in how an `ndb` context is
+created and used in this version:
+
+    - add_flow_exception.
+    - make_context.
+    - make_default_context.
+    - QueueFuture.
+    - ReducedFuture.
+    - SerialQueueFuture.
+    - set_context.
+    - toplevel.
+
+Utils
+=====
+
+The previous version of `ndb` included an ``ndb.utils`` module, which defined
+a number of methods that were mostly used internally. Some of those have been
+made obsolete by new Python 3 features, while others have been discarded due
+to implementation differences in the new `ndb`.
+
+Possibly the most used utility from this module outside of `ndb` code, is the
+``positional`` decorator, which declares that only the first `n` arguments of
+a function or method may be positional. Python 3 can do this using keyword-only
+arguments. What used to be written as::
+
+    @utils.positional(2)
+    def function1(arg1, arg2, arg3=None, arg4=None)
+        pass
+
+Will be written like this in the new version::
+
+    def function1(arg1, arg2, *, arg3=None, arg4=None)
+        pass
+
+Exceptions
+==========
+
+App Engine's legacy exceptions are no longer available, but `ndb` provides
+shims for most of them, which can be imported from the `ndb.exceptions`
+package, like this::
+
+    from ndb.exceptioms import BadRequestError, BadArgumentError
+
+Datastore API
+=============
+
+There are many differences bewteen the current Datastore API and the legacy App
+Engine Datastore. In most cases, where the public API was generally used, this
+should not be a problem. However, if you relied in your code on the private
+Datastore API, the code that does this will probably need to be rewritten.
+Specifically, any function or method that dealt directly with protocol buffers
+will no longer work. The Datastore `.protobuf` definitions have changed
+significantly from the public API used by App Engine to the current published
+API. Additionally, this version of NDB mostly delegates to
+`google.cloud.datastore` for parsing data returned by RPCs, which is a
+significant internal refactoring.
+
+Default Namespace
+=================
+
+In the previous version, ``google.appengine.api.namespacemanager`` was used
+to determine the default namespace when not passed in to constructors which
+require it, like ``Key``. In this version, the client class can be instantiated
+with a namespace, which will be used as the default whenever it's not included
+in the constuctor or method arguments that expect a namespace::
+
+    from google.cloud import ndb
+
+    client=ndb.Client(namespace="my namespace")
+    
+    with context as client.context():
+        key = ndb.Key("SomeKind", "SomeId")
+
+In this example, the key will be created under the namespace `my namespace`,
+because that's the namespace passed in when setting up the client.
+
+Django Middleware
+=================
```

Comment:
So, a couple things here:

1) A context is thread local. Recommended practice is one context per request. So you probably want to create the client in ``__init__`` but create the context in ``__call__``.

2) ``__call__`` looks like a great place to use a ``with`` statement.

Maybe something more like:

~~~~
    from google.cloud import ndb

    class NDBMiddleware(object):
        def __init__(self, get_response):
            self.get_response = get_response
            self.client = ndb.Client()

        def __call__(self, request):
            context = client.context()
            request.ndb_context = context
            with context.use():
                response = self.get_response(request)
            return response
~~~~

Reply:
If it is any help, putting this in my Django middleware, I am receiving `AttributeError: '_GeneratorContextManager' object has no attribute 'use'`. Seems to be working without the `use` but I don't have enough "context" to know if that will break something else.

Putting the context in the `__init__` definitely broke things for the second request that uses `ndb`.

Reply:
Thanks a lot, I'm going to test this more and make it work before the merge.

Reply:
Right, I forgot that ``Client.context`` calls ``Context.use`` for you. I will correct my snippet.